src/InteractiveWindow/Editor/InteractiveWindow.UIThreadOnly.cs

```diff
@@ -479,12 +495,51 @@ public void InsertCode(string text)
                 }
                 else
                 {
+                    InsertText(text);
+                }
+            }
+
+            private void InsertText(string text)
+            {
+                using (var transaction = UndoHistory_TestOnly.CreateTransaction(InteractiveWindowResources.TypeChar))
+                {
+                    var selection = TextView.Selection;
+                    var caretPosition = TextView.Caret.Position.BufferPosition;
                     if (!TextView.Selection.IsEmpty)
                     {
-                        CutOrDeleteSelection(isCut: false);
+                        if (!IsSelectionInsideCurrentSubmission())
+                        {
+                            return;
+                        }
+
+                        DeleteSelection();
+
+                        if (selection.Mode == TextSelectionMode.Box)
+                        {
+                            ReduceBoxSelectionToEditableBox(isDelete: true);
+                        }
+                        else
+                        {
+                            selection.Clear();
+                            MoveCaretToClosestEditableBuffer();
+                        }
+                    }
+                    else if (IsInActivePrompt(caretPosition))
+                    {
+                        MoveCaretToClosestEditableBuffer();
+                    }
+                    else if (MapToEditableBuffer(caretPosition) == null)
+                    {
+                        return;
                     }
 
                     EditorOperations.InsertText(text);
+
+                    // don't record undo histroy for standard-input
+                    if (!ReadingStandardInput)
```

Comment:
You mean `transaction.Complete()`? It mark the transaction as complete and put it into the history stack. `transaction.Dispose()` would not save to the stack.